From: Thomas Gleixner <tglx@linutronix.de>
Date: Tue, 26 Sep 2023 13:03:52 +0000
Subject: [PATCH 100/108] printk: nbcon: Add printer thread wakeups
Origin: https://www.kernel.org/pub/linux/kernel/projects/rt/6.6/older/patches-6.6-rt12.tar.xz

Add a function to wakeup the printer threads. Use the new function
when:

  - records are added to the printk ringbuffer
  - consoles are started
  - consoles are resumed
  - exit from an atomic printing section
  - triggered via printk_trigger_flush()

The actual waking is performed via irq_work so that the wakeup can
be triggered from any context.

Co-developed-by: John Ogness <john.ogness@linutronix.de>
Signed-off-by: John Ogness <john.ogness@linutronix.de>
Signed-off-by: Thomas Gleixner (Intel) <tglx@linutronix.de>
Signed-off-by: Sebastian Andrzej Siewior <bigeasy@linutronix.de>
---
 include/linux/console.h  |    3 ++
 kernel/printk/internal.h |    1 
 kernel/printk/nbcon.c    |   62 +++++++++++++++++++++++++++++++++++++++++++++++
 kernel/printk/printk.c   |    9 ++++++
 4 files changed, 74 insertions(+), 1 deletion(-)

--- a/include/linux/console.h
+++ b/include/linux/console.h
@@ -16,6 +16,7 @@
 
 #include <linux/atomic.h>
 #include <linux/bits.h>
+#include <linux/irq_work.h>
 #include <linux/rculist.h>
 #include <linux/rcuwait.h>
 #include <linux/types.h>
@@ -305,6 +306,7 @@ struct nbcon_write_context {
  * @locked_port:	True, if the port lock is locked by nbcon
  * @kthread:		Printer kthread for this console
  * @rcuwait:		RCU-safe wait object for @kthread waking
+ * @irq_work:		Defer @kthread waking to IRQ work context
  */
 struct console {
 	char			name[16];
@@ -337,6 +339,7 @@ struct console {
 	bool			locked_port;
 	struct task_struct	*kthread;
 	struct rcuwait		rcuwait;
+	struct irq_work		irq_work;
 };
 
 #ifdef CONFIG_LOCKDEP
--- a/kernel/printk/internal.h
+++ b/kernel/printk/internal.h
@@ -79,6 +79,7 @@ void nbcon_init(struct console *con);
 void nbcon_free(struct console *con);
 bool nbcon_console_emit_next_record(struct console *con);
 void nbcon_kthread_create(struct console *con);
+void nbcon_wake_threads(void);
 
 /*
  * Check if the given console is currently capable and allowed to print
--- a/kernel/printk/nbcon.c
+++ b/kernel/printk/nbcon.c
@@ -1106,6 +1106,61 @@ static int nbcon_kthread_func(void *__co
 }
 
 /**
+ * nbcon_irq_work - irq work to wake printk thread
+ * @irq_work:	The irq work to operate on
+ */
+static void nbcon_irq_work(struct irq_work *irq_work)
+{
+	struct console *con = container_of(irq_work, struct console, irq_work);
+
+	nbcon_kthread_wake(con);
+}
+
+static inline bool rcuwait_has_sleeper(struct rcuwait *w)
+{
+	bool has_sleeper;
+
+	rcu_read_lock();
+	/*
+	 * Guarantee any new records can be seen by tasks preparing to wait
+	 * before this context checks if the rcuwait is empty.
+	 *
+	 * This full memory barrier pairs with the full memory barrier within
+	 * set_current_state() of ___rcuwait_wait_event(), which is called
+	 * after prepare_to_rcuwait() adds the waiter but before it has
+	 * checked the wait condition.
+	 *
+	 * This pairs with nbcon_kthread_func:A.
+	 */
+	smp_mb(); /* LMM(rcuwait_has_sleeper:A) */
+	has_sleeper = !!rcu_dereference(w->task);
+	rcu_read_unlock();
+
+	return has_sleeper;
+}
+
+/**
+ * nbcon_wake_threads - Wake up printing threads using irq_work
+ */
+void nbcon_wake_threads(void)
+{
+	struct console *con;
+	int cookie;
+
+	cookie = console_srcu_read_lock();
+	for_each_console_srcu(con) {
+		/*
+		 * Only schedule irq_work if the printing thread is
+		 * actively waiting. If not waiting, the thread will
+		 * notice by itself that it has work to do.
+		 */
+		if (con->kthread && rcuwait_has_sleeper(&con->rcuwait))
+			irq_work_queue(&con->irq_work);
+	}
+	console_srcu_read_unlock(cookie);
+}
+
+/**
  * struct nbcon_cpu_state - Per CPU printk context state
  * @prio:	The current context priority level
  * @nesting:	Per priority nest counter
@@ -1323,6 +1378,7 @@ enum nbcon_prio nbcon_atomic_enter(enum
 void nbcon_atomic_exit(enum nbcon_prio prio, enum nbcon_prio prev_prio)
 {
 	struct nbcon_cpu_state *cpu_state;
+	u64 next_seq = prb_next_seq(prb);
 
 	__nbcon_atomic_flush_all(false);
 
@@ -1343,6 +1399,11 @@ void nbcon_atomic_exit(enum nbcon_prio p
 	 */
 	cpu_state->prio = prev_prio;
 
+	if (cpu_state->nesting[cpu_state->prio] == 0 &&
+	    prb_read_valid(prb, next_seq, NULL)) {
+		nbcon_wake_threads();
+	}
+
 	preempt_enable();
 }
 
@@ -1466,6 +1527,7 @@ void nbcon_init(struct console *con)
 	BUG_ON(!con->pbufs);
 
 	rcuwait_init(&con->rcuwait);
+	init_irq_work(&con->irq_work, nbcon_irq_work);
 	nbcon_seq_force(con, con->seq);
 	nbcon_state_set(con, &state);
 	nbcon_kthread_create(con);
--- a/kernel/printk/printk.c
+++ b/kernel/printk/printk.c
@@ -2286,6 +2286,8 @@ asmlinkage int vprintk_emit(int facility
 
 	printed_len = vprintk_store(facility, level, dev_info, fmt, args);
 
+	nbcon_wake_threads();
+
 	/* If called from the scheduler, we can not call up(). */
 	if (!IS_ENABLED(CONFIG_PREEMPT_RT) && !in_sched) {
 		/*
@@ -2307,7 +2309,6 @@ asmlinkage int vprintk_emit(int facility
 		preempt_enable();
 	}
 
-
 	if (in_sched)
 		defer_console_output();
 	else
@@ -2577,6 +2578,11 @@ void resume_console(void)
 	 */
 	synchronize_srcu(&console_srcu);
 
+	/*
+	 * Since this runs in task context, wake the threaded printers
+	 * directly rather than scheduling irq_work to do it.
+	 */
+
 	cookie = console_srcu_read_lock();
 	for_each_console_srcu(con) {
 		flags = console_srcu_read_flags(con);
@@ -3969,6 +3975,7 @@ void printk_trigger_flush(void)
 	nbcon_atomic_flush_all();
 	preempt_enable();
 
+	nbcon_wake_threads();
 	defer_console_output();
 }
 
